File: src/main/java/fr/adrienbrault/idea/symfony2plugin/twig/action/TwigTranslationGeneratorAction.java
package fr.adrienbrault.idea.symfony2plugin.twig.action;

import com.intellij.codeInsight.CodeInsightActionHandler;
import com.intellij.codeInsight.actions.CodeInsightAction;
import com.intellij.codeInsight.lookup.LookupElement;
import com.intellij.openapi.command.WriteCommandAction;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.ui.popup.JBPopupFactory;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;
import com.intellij.psi.impl.source.html.HtmlFileImpl;
import com.jetbrains.php.completion.insert.PhpInsertHandlerUtil;
import com.jetbrains.twig.TwigFile;
import fr.adrienbrault.idea.symfony2plugin.Symfony2ProjectComponent;
import fr.adrienbrault.idea.symfony2plugin.templating.util.TwigUtil;
import fr.adrienbrault.idea.symfony2plugin.translation.dict.TranslationUtil;
import org.apache.commons.lang.StringUtils;
import org.jetbrains.annotations.NotNull;

import java.util.List;
import java.util.stream.Collectors;

/**
 * @author Daniel Espendiller <daniel@espendiller.net>
 */
public class TwigTranslationGeneratorAction extends CodeInsightAction {
    @Override
    protected boolean isValidForFile(@NotNull Project project, @NotNull Editor editor, @NotNull PsiFile file) {
        return Symfony2ProjectComponent.isEnabled(project) && (
            file instanceof TwigFile
            || (file instanceof HtmlFileImpl && file.getName().toLowerCase().endsWith(".twig"))
            || TwigUtil.getInjectedTwigElement(file, editor) != null
        );
    }

    @NotNull
    @Override
    protected CodeInsightActionHandler getHandler() {
        return new MyCodeInsightActionHandler();
    }

    private static class MyCodeInsightActionHandler implements CodeInsightActionHandler {
        @Override
        public void invoke(@NotNull Project project, @NotNull Editor editor, @NotNull PsiFile psiFile) {
            PsiElement psiElement = TwigUtil.getInjectedTwigElement(psiFile, editor);
            if(psiElement == null) {
                return;
            }

            PsiFile containingFile = psiElement.getContainingFile();
            if(!(containingFile instanceof TwigFile)) {
                return;
            }

            PsiElement element = TwigUtil.getElementOnTwigViewProvider(psiElement);
            TwigUtil.DomainScope twigFileDomainScope = TwigUtil.getTwigFileDomainScope(element != null ? element : psiElement);

            final String defaultDomain = twigFileDomainScope.getDefaultDomain();
            final String domain = twigFileDomainScope.getDomain();

            final List<String> list = TranslationUtil.getTranslationLookupElementsOnDomain(project, domain)
                .stream()
                .map(LookupElement::getLookupString)
                .sorted()
                .collect(Collectors.toList());

            JBPopupFactory.getInstance().createPopupChooserBuilder(list)
                .setTitle(String.format("Symfony: Translations \"%s\"", StringUtils.abbreviate(domain, 20)))
                .setItemChosenCallback(selectedValue -> WriteCommandAction.runWriteCommandAction(editor.getProject(), String.format("Symfony: Add Translation \"%s\"", StringUtils.abbreviate(selectedValue, 20)), null, () -> {
                    String s;

                    if (!domain.equals(defaultDomain)) {
                        s = String.format("{{ '%s'|trans({}, '%s') }}", selectedValue, domain);
                    } else {
                        s = String.format("{{ '%s'|trans }}", selectedValue);
                    }

                    PhpInsertHandlerUtil.insertStringAtCaret(editor, s);
                }))
                .createPopup()
                .showInBestPositionFor(editor);
        }
    }
}

File: src/main/kotlin/fr/adrienbrault/idea/symfony2plugin/twig/action/TwigTranslationGeneratorAction.kt
package fr.adrienbrault.idea.symfony2plugin.twig.action

import com.intellij.codeInsight.CodeInsightActionHandler
import com.intellij.codeInsight.actions.CodeInsightAction
import com.intellij.openapi.command.WriteCommandAction
import com.intellij.openapi.editor.Editor
import com.intellij.openapi.project.Project
import com.intellij.openapi.ui.popup.JBPopupFactory
import com.intellij.psi.PsiFile
import com.intellij.psi.impl.source.html.HtmlFileImpl
import com.jetbrains.php.completion.insert.PhpInsertHandlerUtil
import com.jetbrains.twig.TwigFile
import fr.adrienbrault.idea.symfony2plugin.Symfony2ProjectComponent
import fr.adrienbrault.idea.symfony2plugin.templating.util.TwigUtil
import fr.adrienbrault.idea.symfony2plugin.translation.dict.TranslationUtil
import org.apache.commons.lang3.StringUtils

/**
 * @author Daniel Espendiller <daniel@espendiller.net>
 */
class TwigTranslationGeneratorAction : CodeInsightAction() {
    override fun isValidForFile(project: Project, editor: Editor, file: PsiFile) =
        Symfony2ProjectComponent.isEnabled(project) &&
            (file is TwigFile ||
                file is HtmlFileImpl && file.name.endsWith(".twig", ignoreCase = true) ||
                TwigUtil.getInjectedTwigElement(file, editor) != null)

    override fun getHandler(): CodeInsightActionHandler = MyCodeInsightActionHandler()

    private class MyCodeInsightActionHandler : CodeInsightActionHandler {
        override fun invoke(project: Project, editor: Editor, psiFile: PsiFile) {
            val psiElement = TwigUtil.getInjectedTwigElement(psiFile, editor) ?: return

            val containingFile = psiElement.containingFile
            if (containingFile !is TwigFile) {
                return
            }

            val element = TwigUtil.getElementOnTwigViewProvider(psiElement)
            val twigFileDomainScope = TwigUtil.getTwigFileDomainScope(element ?: psiElement)

            val defaultDomain = twigFileDomainScope.defaultDomain
            val domain = twigFileDomainScope.domain

            val list = TranslationUtil.getTranslationLookupElementsOnDomain(project, domain)
                .map { it.lookupString }
                .sorted()

            JBPopupFactory.getInstance().createPopupChooserBuilder(list)
                .setTitle("Symfony: Translations \"${StringUtils.abbreviate(domain, 20)}\"")
                .setItemChosenCallback { selectedValue ->
                    WriteCommandAction.runWriteCommandAction(
                        editor.project,
                        "Symfony: Add Translation \"${StringUtils.abbreviate(selectedValue, 20)}\"",
                        null,
                        {
                            PhpInsertHandlerUtil.insertStringAtCaret(
                                editor,
                                createTranslationSnippet(selectedValue, defaultDomain, domain)
                            )
                        }
                    )
                }
                .createPopup()
                .showInBestPositionFor(editor)
        }
    }
}

fun createTranslationSnippet(translationKey: String, defaultDomain: String, domain: String) =
    if (domain != defaultDomain) {
        "{{ '$translationKey'|trans({}, '$domain') }}"
    } else {
        "{{ '$translationKey'|trans }}"
    }
